### PR TITLE
Add tests for discharge and discharge_all.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

File: macaroonbakery/bakery.py [hunk 1]
```diff
@@ -157,12 +160,64 @@ def _extract_urls(response):
...
-    def __init__(self, version, public_key):
-        '''
-        @param version holds latest the bakery protocol version supported
-        by the discharger.
-        @param public_key holds the public nacl key of the third party.
+class Bakery(object):
+    '''Convenience class that contains both an Oven and a Checker.
+    '''
+    def __init__(self, location=None, locator=None, ops_store=None, key=None,
+                 identity_client=None, checker=None, root_key_store=None,
```
Comment (rogpeppe) [inline macaroonbakery/bakery.py]: checker=checkers.Checker() ?

File: macaroonbakery/discharge.py [hunk 2]
```diff
@@ -76,18 +81,19 @@ def check_third_party_caveat(self, ctx, info):
         :param ctx (AuthContext)
         :param info (ThirdPartyCaveatInfo) holds the information decoded from
         a third party caveat id
-        :return: If the caveat is valid, it returns a None error and optionally
-         a slice of extra caveats that will be added to the discharge macaroon
+        :return: If the caveat is validi might return optionally
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: :return: An array of extra caveats to be added to the discharge macaroon.

(leaving the description of the returned exceptions for the start) ?

File: macaroonbakery/discharge.py [hunk 3]
```diff
@@ -76,18 +81,19 @@ def check_third_party_caveat(self, ctx, info):
         :param ctx (AuthContext)
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: If the caveat kind was not recognised, the checker should
    raise a CaveatNotRecognized exception; if the check failed,
    it should raise a ThirdPartyCaveatCheckFailed exception.


?

File: macaroonbakery/discharge.py [hunk 4]
```diff
@@ -113,7 +119,7 @@ def discharge(ctx, id, caveat, key, checker, locator):
     party caveats returned by the Checker.
     '''
     caveat_id_prefix = []
-    if caveat is None:
+    if caveat is None or caveat == b'':
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: When does this happen? This doesn't match the doc comment description - why do we need to check against the empty string too?

File: macaroonbakery/discharge.py [hunk 5]
```diff
@@ -129,7 +135,15 @@ def discharge(ctx, id, caveat, key, checker, locator):
     cond, arg = checkers.parse_caveat(cav_info.condition)
 
     if cond == checkers.COND_NEED_DECLARED:
-        cav_info.condition = arg.encode('utf-8')
+        cav_info = ThirdPartyCaveatInfo(
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: cav_info = cav_info._replace(condition=arg.encode('utf-8'))

?

File: macaroonbakery/discharge.py [hunk 6]
```diff
@@ -160,11 +174,19 @@ def _check_need_declared(ctx, cav_info, checker):
                              'attribute')
     if len(need_declared) == 0:
         raise ValueError('need-declared caveat with no required attributes')
-    cav_info.condition = arg[i+1:].encode('utf-8')
-    caveats, err = checker.check_third_party_caveat(ctx, cav_info)
+    cav_info = ThirdPartyCaveatInfo(
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: cav_info = cav_info._replace(condition=arg[i+1:].encode('utf-8'))

?

File: macaroonbakery/macaroon.py [hunk 7]
```diff
@@ -226,7 +239,31 @@ def macaroon_version(bakery_version):
     return pymacaroons.MACAROON_V2
 
 
-def parse_local_location(loc):
+class ThirdPartyLocator(object):
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: This doesn't seem quite right to me. In Go, the ThirdPartyLocator is the interface type, so in Python I'd expect it to be an abstract base class. This type seems the equivalent of ThirdPartyStore - an in-memory implementation of ThirdPartyLocator.

File: macaroonbakery/macaroon.py [hunk 8]
```diff
@@ -226,7 +239,31 @@ def macaroon_version(bakery_version):
...
+        @return: string
+        '''
+        return self._store.get(loc)
+
+    def add_info(self, loc, info):
+        '''Associates the given information with the given location.
+
+        It will ignore any trailing slash.
+        '''
+        self._store[loc.rstrip('\\')] = info
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: This strips the wrong kind of slash - we're interested in removing the training / from URLs here.

File: macaroonbakery/macaroon.py [hunk 9]
```diff
@@ -241,40 +278,18 @@ def parse_local_location(loc):
     @return a tuple of location and if the location is local.
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: Looks like this is no accurate.

File: macaroonbakery/tests/test_macaroon.py [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
-
 from unittest import TestCase
 
 import nacl.utils
 
 from macaroonbakery import LATEST_BAKERY_VERSION, BAKERY_V1, macaroon, codec
-from macaroonbakery.bakery import ThirdPartyInfo
 from macaroonbakery.macaroon import ThirdPartyLocator
-from macaroonbakery.third_party import legacy_namespace, ThirdPartyCaveatInfo
+from macaroonbakery.third_party import (
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_macaroon.py]: It looks to me as if the test_add_third_party_caveat method in this file still needs to be updated, or is that going to be a separate PR?

File: macaroonbakery/third_party.py [hunk 11]
```diff
@@ -55,3 +55,14 @@ def __eq__(self, other):
             self.version == other.version and
             self.ns == other.ns
         )
+
+
+class ThirdPartyInfo:
```
Comment (rogpeppe) [inline macaroonbakery/third_party.py]: Shouldn't this inherit from a namedtuple class?

File: macaroonbakery/checkers/checkers.py [hunk 12]
```diff
@@ -49,10 +49,10 @@ class Checker(FirstPartyCaveatChecker):
     ''' Holds a set of checkers for first party caveats.
     '''
 
-    def __init__(self, ns=None, include_std_checkers=True):
-        if ns is None:
-            ns = Namespace()
-        self._ns = ns
+    def __init__(self, namespace=None, include_std_checkers=True):
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checkers/checkers.py]: namespace has a state so probably not.

File: macaroonbakery/discharge.py (hunk 5 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: a namedtuple is immutable unless you use protected method like this one, I prefer to create a new instance.

File: macaroonbakery/tests/test_macaroon.py (hunk 10 above)
Comment (fabricematrat) [reply macaroonbakery/tests/test_macaroon.py]: yep, next PR

File: macaroonbakery/checkers/namespace.py [hunk 13]
```diff
@@ -141,6 +141,8 @@ def deserialize_namespace(data):
     '''
     if isinstance(data, bytes):
         data = data.decode('utf-8')
+    if data == '':
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checkers/namespace.py]: done.

File: macaroonbakery/checkers/namespace.py [hunk 14]
```diff
@@ -141,6 +141,8 @@ def deserialize_namespace(data):
...
+        return Namespace()
     kvs = data.split(' ')
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checkers/namespace.py]: done.

File: macaroonbakery/codec.py [hunk 15]
```diff
@@ -168,7 +167,7 @@ def decode_caveat(key, caveat):
                 'caveat id payload not provided for caveat id {}'.format(
                     caveat))
         return _decode_caveat_v2_v3(first_as_int, key, caveat)
-    raise NotImplementedError('only bakery v1 supported')
+    raise ValueError('unknow version for caveat')
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/codec.py]: done.

File: macaroonbakery/discharge.py (hunk 3 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: done.

File: macaroonbakery/discharge.py (hunk 2 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: done.

File: macaroonbakery/discharge.py (hunk 4 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: my fault

File: macaroonbakery/discharge.py (hunk 6 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: same

File: macaroonbakery/macaroon.py (hunk 7 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: done.

File: macaroonbakery/macaroon.py (hunk 8 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: done.

File: macaroonbakery/macaroon.py (hunk 9 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: done.

File: macaroonbakery/third_party.py (hunk 11 above)
Comment (fabricematrat) [reply macaroonbakery/third_party.py]: done

File: macaroonbakery/discharge.py (hunk 5 above)
Comment (rogpeppe) [reply macaroonbakery/discharge.py]: the _replace method does return a new instance - that's the point. Using _replace is both cleaner and less error-prone IMHO.

File: macaroonbakery/checkers/namespace.py [hunk 16]
```diff
@@ -141,11 +141,23 @@ def deserialize_namespace(data):
...
-        uri_to_prefix[k] = v
+        i = kv.rfind(':')
+        if i == -1:
+            raise ValueError('no colon in namespace field {}'.format(kv))
+        uri, prefix = kv[0:i], kv[i + 1:]
+        if not is_valid_schema_uri(uri):
+            # Currently this can't happen because the only invalid URIs
+            # are those which contain a space
+            raise ValueError(
+                'invalid URI {} in namespace field {}'.format(uri, kv))
```
Comment (rogpeppe) [inline macaroonbakery/checkers/namespace.py]: Perhaps use repr(uri) and repr(kv) (and below) so the messages are more obvious if the strings are empty or otherwise ambiguous?

File: macaroonbakery/macaroon.py [hunk 17]
```diff
@@ -242,25 +244,32 @@ def macaroon_version(bakery_version):
...
-
         @param loc string
         @return: string
         '''
-        return self._store.get(loc)
+        raise NotImplementedError('third_party_info method must be defined in '
+                                  'subclass')
+
+
+class ThirdPartyStore(ThirdPartyLocator):
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: doc comments please.
